Userland/Applications/Browser/BookmarksBarWidget.cpp

```diff
@@ -321,22 +343,41 @@ bool BookmarksBarWidget::edit_bookmark(DeprecatedString const& url)
         auto item_url = model()->index(item_index, 1).data().to_deprecated_string();
 
         if (item_url == url) {
-            auto values = BookmarkEditor::edit_bookmark(window(), item_title, item_url);
-            bool item_replaced = false;
+            auto values = BookmarkEditor::edit_bookmark(window(), item_title, item_url, PerformEditOn::ExistingBookmark);
+            return store_existing_bookmark_edit(item_index, values);
+        }
+    }
+
+    return false;
+}
 
-            if (!values.is_empty()) {
-                auto& json_model = *static_cast<GUI::JsonArrayModel*>(model());
-                item_replaced = json_model.set(item_index, move(values));
+bool BookmarksBarWidget::store_new_bookmark_edit(DeprecatedString const& url_before_editor_dialog, Vector<JsonValue>& values)
+{
+    if (values.is_empty())
+        return false;
 
-                if (item_replaced)
-                    json_model.store();
-            }
+    for (int item_index = 0; item_index < model()->row_count(); ++item_index) {
+        auto item_url = model()->index(item_index, 1).data().to_deprecated_string();
 
-            return item_replaced;
-        }
+        if (item_url == url_before_editor_dialog)
+            return store_existing_bookmark_edit(item_index, values);
     }
 
     return false;
 }
 
+bool BookmarksBarWidget::store_existing_bookmark_edit(int bookmark_index, Vector<JsonValue>& values)
```

Comment:
stylistically itʼs somewhat different than `store_new_bookmark` (i like the code the better).

it seems you could even have a common function out of those, as differ only the model mutating function which could be placed in a lambda… for me that could be a nice thing. 🤔

still itʼs good as is for me, so thatʼs up to you whether you want change it or leave be (resolve this comment if so). 🙃

Reply:
Thank you for your input! I did not follow your common function idea with lamdas as I wasn't sure how to best approach it.  However, you did help me realize that I could do away with my `store_new_bookmark_edit()` helper method and just use `edit_bookmark()` in `add_bookmark()` since its return value will tell me if the bookmark has changed (i.e. if we pressed the ok button in the bookmark editor and we were able to successfully store it). I also renamed the `store_existing_bookmark_edit()` helper function to `store_bookmark_at()` in order to better describe what to expect it arguments to be.

I will resolve this comment, but if you believe these changes can be further improved please let me know!

Reply:
well… what i wanted to point out is that `store_new_bookmark` and `store_bookmark_at` differ only in one call – used exactly in the same context – i.e. `json_model.add(move(values))` and `json_model.set(bookmark_index, move(values))` resp. you could extract common logic to a function accepting `values` and a lambda with above snippets passed in the call place.

this would make things über-dry, so it is not a big thing as i wrote before and itʼs completely acceptable as-is. 🙂

Reply:
Now I understand what you mean. So we could create a single helper method called `update_model()` and we could say it has the following definition:
```cpp
bool BookmarksBarWidget::update_model(Vector<JsonValue>& values, Function<bool(GUI::JsonArrayModel& model, Vector<JsonValue>& values)> perform_model_change)
{
    bool was_model_changed = false;

    if (!values.is_empty()) {
        auto& json_model = *static_cast<GUI::JsonArrayModel*>(model());
        was_model_changed = perform_model_change(json_model, values);

        if (was_model_changed )
            json_model.store();
    }

    return was_model_changed;
}
```
If it were to add a bookmark it would be like so:
```cpp
update_model(values, [](auto& model, auto& values) {
    return model.add(move(values));
});
```

If we were to set a bookmark after it has been edited:
```cpp
update_model(values, [item_index](auto& model, auto& values) {
    return model.set(item_index, move(values));
});
```

I believe the main tradeoff is readability but since this is just a helper method I don't really see a problem with that. I'll go ahead and implement this.

Reply:
Added the new changes. The only thing that's different from what is shown above is the lambda takes an rvalue reference of `Vector<JsonValue> values` instead of an lvalue reference.

Reply:
🥹